File: test-case/tree/mirror/Mirror.kt
import soda.kotlin.leetcode.*
import soda.kotlin.unittest.*

class Solution {
    fun mirror(root: TreeNode?): TreeNode? {
        if (root == null) {
            return null
        }
        mirror(root.left)
        mirror(root.right)
        val temp = root.left
        root.left = root.right
        root.right = temp
        return root
    }
}

class Mirror : (String) -> String {
    override fun invoke(text: String): String {
        // * setup with your method
        val work = GenericTestWork.create(Solution()::mirror)

        // * by method that has not return value
        // val work = GenericTestWork.createVoid(Solution.METHOD_WITHOUT_RETURN)
        // * by class of data struct
        // val work = GenericTestWork.forStruct(STRUCT::class)

        // * setup validator
        // work.validator = (T, T) -> Boolean
        work.compareSerial = true
        return work.run(text)
    }

}

fun main() {
    print(Mirror()(Utils.fromStdin()))
}

File: test-case/tree/mirror/mirror.php
<?php
require_once 'Soda/DS/index.php';
require_once 'Soda/Leetcode/index.php';
require_once 'Soda/Unittest/index.php';

use Soda\DS;
use Soda\Leetcode\ListNode;
use Soda\Leetcode\NestedInteger;
use Soda\Leetcode\TreeNode;
use Soda\Unittest\TestWork;
use Soda\Unittest\Utils;
use Soda\Unittest\Validators;

class Solution {

    /**
     * @param TreeNode $root
     * @return TreeNode 
     */
    function mirror($root) {
        if (!$root) {
            return $root;
        }
        $this->mirror($root->left);
        $this->mirror($root->right);
        [$root->left, $root->right] = [$root->right, $root->left];
        return $root;
    }
}

$work = TestWork::create(new Solution(), 'mirror', $argv[0]);
// $work->validator = fn($e, $r) => <boolean result>
$work->compareSerial = true;
echo $work->run(Utils::fromStdin());
